improve aws utils and database validation

GavinFrazar opened the pull request:
This PR is kind of a grab-bag of changes that I want to backport.

It's all work I pulled out of changes I'm preparing for https://github.com/gravitational/teleport/issues/21872

May be best reviewed commit-by-commit, since I organized the commits by purpose with explanations.

Edit: seems this broke some app access tests. Need to check if we allow a short role name like "test-role" instead of requiring the full role arn like "arn:aws:iam::123456789012::roles/test-role". I'm not sure we support that for app access, so the tests themselves could just need fixing.

Edit2: okay we do allow the short name. This regression is from changes in `FilterAWSRoles` which is used by `tsh app` when searching the .tsh profile trying to find an AWS role ARN. The profile's role ARNs should always be saved fully specified though. So I think this is just a test bug. Will fix tomorrow.

smallinsky commented on lib/utils/aws/aws.go:
```diff
@@ -348,37 +344,87 @@ func isJSON(contentType string) bool {
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+		// don't check check for empty accountID - callers do not always pass an account ID,
```
```suggestion
		// don't check for empty accountID - callers do not always pass an account ID,
```

smallinsky commented on lib/utils/aws/aws.go:
```diff
@@ -348,37 +344,87 @@ func isJSON(contentType string) bool {
 
 // BuildRoleARN constructs a string AWS ARN from a username, region, and account ID.
-func BuildRoleARN(username, region, accountID string) string {
+func BuildRoleARN(username, region, accountID string) (string, error) {
```
Could you add a bit more comment to this function and explain why the username values can be either full ARN only user name or `role` name ?

smallinsky commented on lib/srv/db/common/auth.go:
```diff
@@ -816,32 +816,11 @@ func matchAzureResourceName(resourceID, name string) bool {
-			Resource:  resource,
-		}.String(), nil
 	}
+	return awsutils.BuildRoleARN(username, aws.Region, aws.AccountID)
```
nit: 
```suggestion
	val, err  = awsutils.BuildRoleARN(username, aws.Region, aws.AccountID)
	return val, trace.Wrap(err)
```

greedy52 commented on lib/utils/aws/aws.go:
```diff
@@ -348,37 +344,87 @@ func isJSON(contentType string) bool {
+// Example role ARN: arn:aws:iam::123456789012:role/some-role-name
+func checkRoleARN(parsed *arn.ARN) error {
+	parts := strings.Split(parsed.Resource, "/")
+	if parts[0] != "role" || parsed.Service != "iam" {
```
nit:
```suggestion
	if parts[0] != "role" || parsed.Service != iam.ServiceName {
```

greedy52 commented on lib/utils/aws/aws_test.go:
```diff
@@ -255,3 +255,140 @@ func TestValidateRoleARNAndExtractRoleName(t *testing.T) {
+		"iam role arn is missing role name": {
+			arn:             "arn:aws:iam::123456789012:role",
+			wantErrContains: "missing AWS IAM role name",
+		},
```
do you mind adding a test for a role with multiple levels of `/`
for example:
```
arn:aws:iam::123456789012:role/aws-reserved/sso.amazonaws.com/us-west-2/AWSReservedSSO_AWSPowerUserAccess_xxxxxxxxx
arn:aws:iam::123456789012:role/aws-service-role/redshift.amazonaws.com/AWSServiceRoleForRedshift
```

greedy52 commented on lib/utils/aws/aws_test.go:
```diff
@@ -255,3 +255,140 @@ func TestValidateRoleARNAndExtractRoleName(t *testing.T) {
+	for name, tt := range tests {
+		tt := tt
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
```
nit: IMHO these quick tests don't benefit much from parallel tests.

GavinFrazar replied:
👍 good point, I parallelized them since they were trivial but it makes no difference for runtime. will remove

greedy52 reviewed (approved):
for the test failure, `makeUserWithAWSRole` was using an invalid role ARN. 

I love the idea filter out the bad ones. just from UX perspective, I am wondering if we should show a warning if we detected some bad ARNs during `FilterAWSRoles`.

GavinFrazar committed:
improve aws utils and database validation

* log a warning when an invalid AWS IAM role ARN is filtered out
* add tests for nested role arn path and remove superfluous test parallelism
* add extra godoc info for BuildRoleARN
* improve mock IAM error messages
  - Failing tests' error messages were not specific enough to know what
    happened.
* refactor code to call GetAWS() once
  - Avoids constantly fetching a full copy of the AWS metadata.
  - Reduces verbosity of code.
* reuse MetadataFromRedshiftCluster in meta fetcher
  - Avoids repeating the same logic inside the fetcher.
* improve database-access DynamoDB config checking
  - Don't early return from database CheckAndSetDefaults so AWS account ID
    is checked properly, and any other future checks added will apply to
    DynamoDB as well.
* add aws utils for parsing AWS IAM roles
  - ParseRoleARN will parse an AWS ARN and verify that it is a valid
    AWS IAM Role resource
  - CheckARNPartitionAndAccount is a helper that verifies an ARN partition
    and account ID matches the caller's expectations.
* use cmp.Diff instead of require.Equal in db tests
  - require.Equal spams a lot of protobuf noise.
  - cmp.Diff yields a precise, readable diff on test failure.
